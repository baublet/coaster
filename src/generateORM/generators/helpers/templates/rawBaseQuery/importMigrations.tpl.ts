import { $$migrationApplicationFunction } from "$$codeOutputFullPath";

testMigrations.push($$migrationApplicationFunction);
