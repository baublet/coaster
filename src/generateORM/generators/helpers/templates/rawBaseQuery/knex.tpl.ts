import { Knex, knex } from "knex";
export type Connection = Knex<any, unknown[]>;
export type Transaction = Knex.Transaction<any, any>;
export type ConnectionOrTransaction = Connection | Transaction;
