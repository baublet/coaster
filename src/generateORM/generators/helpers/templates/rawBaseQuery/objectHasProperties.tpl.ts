function objectHasProperties(obj: Record<string, any>, properties: string[]): boolean {
  for(const property of properties) {
    if(!obj.hasOwnProperty(property)) {
      return false;
    }
  }
  return true;
}
