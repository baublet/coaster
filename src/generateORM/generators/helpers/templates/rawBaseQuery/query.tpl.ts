export function $$pluralEntityName<Result = $$entityName[]>(
  connection: ConnectionOrTransaction
) {
  return connection<$$entityName, Result>("$$tableName");
}
