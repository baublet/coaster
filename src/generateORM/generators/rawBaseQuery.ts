import { MetaData, GeneratorResult } from ".";
import { generateNames } from "../../generateNames";
import { RawSchema } from "../drivers";
import { getSchemaAndTablePath } from "./helpers";

/**
 * Generates the lowest-level possible accessor for accessing data in a table
 * using raw database types.
 */
export const rawBaseQuery = (
  schema: RawSchema,
  metaData: MetaData
): GeneratorResult => {
  metaData.setHeader(
    "knex",
    `import knex from "knex";
export type Connection = knex<any, unknown[]>;
export type Transaction = knex.Transaction<any, any>;
export type ConnectionOrTransaction = Connection | Transaction;`
  );

  let code = "";
  let testCode = "";

  for (const table of schema.tables) {
    const entityName = metaData.tableRawEntityNames.get(
      getSchemaAndTablePath(schema.name, table.name)
    );
    const pluralEntityName = generateNames(entityName).pluralPascal;
    code += `export function ${pluralEntityName}<Result = ${entityName}[]>(
  connection: ConnectionOrTransaction
) {
  return connection<${entityName}, Result>("${table.name}");
};\n\n`;
    metaData.rawBaseQueryFunctionNames.set(
      getSchemaAndTablePath(schema.name, table.name),
      pluralEntityName
    );

    if (!metaData.generateTestCode) {
      continue;
    }
    testCode += `import { ${pluralEntityName} } from "${metaData.codeOutputFullPath}"

describe("${pluralEntityName}", () => {
  it("doesn't throw", () => {
    expect(() => ${pluralEntityName}(${metaData.testConnectionVariable})).not.toThrow();
  });
});\n\n`;
  }

  return {
    code,
    testCode,
  };
};
