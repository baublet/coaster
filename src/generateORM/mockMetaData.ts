import { MetaData } from "./generators";

export function getMockMetaData(): MetaData {
  return {
    testConnectionVariable: "connection",
    codeOutputFullPath: "./generated.ts",
    generateTestCode: false,
    entityTableNames: new Map(),
    setHeader: jest.fn(),
    tableEntityNames: new Map(),
    rawBaseQueryFunctionNames: new Map(),
    transformerFunctionNames: {},
    typeAssertionFunctionNames: new Map(),
    typeGuardFunctionNames: new Map(),
    tableRawEntityNames: new Map(),
    namedEntityColumnNames: new Map(),
    namedEntityInputTypeNames: new Map(),
    namedDatabaseEnumNames: new Map(),
    rawDatabaseEnumNames: new Map(),
  };
}
