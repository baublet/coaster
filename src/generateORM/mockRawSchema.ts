import { RawSchema } from "./drivers";

export const getMockRawSchema: () => RawSchema = () => ({
  name: "public",
  flavor: "pg",
  enums: [
    {
      name: "user_account_source",
      values: ["github", "bitbucket"],
    },
  ],
  tables: [
    {
      name: "user",
      primaryKeyColumn: "id",
      primaryKeyType: "string",
      uniqueConstraints: [],
      columns: [
        {
          name: "id",
          columnType: "columnType",
          nullable: false,
          type: "string",
          foreignKeys: [],
          hasDefault: true,
        },
        {
          name: "name",
          comment: "Users don't have to provide their name",
          columnType: "columnType",
          nullable: true,
          type: "string",
          foreignKeys: [],
          hasDefault: false,
        },
        {
          name: "data",
          columnType: "columnType",
          nullable: false,
          type: "AnyJson",
          foreignKeys: [],
          hasDefault: false,
        },
      ],
    },
    {
      name: "user_account",
      comment: "Singular users can have multiple accounts",
      primaryKeyColumn: "id",
      primaryKeyType: "string",
      uniqueConstraints: [],
      columns: [
        {
          name: "id",
          columnType: "columnType",
          nullable: false,
          type: "unknown",
          foreignKeys: [],
          hasDefault: true,
        },
        {
          name: "source",
          comment: "Who sent us this users' data?",
          columnType: "columnType",
          nullable: false,
          type: "enum",
          foreignKeys: [],
          hasDefault: false,
          enumPath: "public.user_account_source",
        },
        {
          name: "source_data",
          comment: "The raw data sent from the source provider",
          columnType: "columnType",
          nullable: false,
          type: "AnyJson",
          foreignKeys: [],
          hasDefault: false,
        },
      ],
    },
  ],
});
