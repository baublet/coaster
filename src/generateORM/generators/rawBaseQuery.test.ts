import { rawBaseQuery } from "./rawBaseQuery";
import { getMockRawSchema } from "../mockRawSchema";
import { getMockMetaData } from "../mockMetaData";

const metaData = getMockMetaData();

metaData.tableRawEntityNames.set("public.user", "RawUser");
metaData.tableRawEntityNames.set("public.user_account", "RawUserAccount");

it("generates a schema", () => {
  expect(rawBaseQuery(getMockRawSchema(), metaData)).toMatchSnapshot();
});

it("generates test code", () => {
  metaData.generateTestCode = true;
  expect(rawBaseQuery(getMockRawSchema(), metaData)).toMatchSnapshot();
});
