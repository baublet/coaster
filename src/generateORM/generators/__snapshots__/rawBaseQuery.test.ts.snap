// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates a schema 1`] = `
Object {
  "code": "export function RawUsers<Result = RawUser[]>(
  connection: ConnectionOrTransaction
) {
  return connection<RawUser, Result>(\\"user\\");
};

export function RawUserAccounts<Result = RawUserAccount[]>(
  connection: ConnectionOrTransaction
) {
  return connection<RawUserAccount, Result>(\\"user_account\\");
};

",
  "testCode": "",
}
`;

exports[`generates test code 1`] = `
Object {
  "code": "export function RawUsers<Result = RawUser[]>(
  connection: ConnectionOrTransaction
) {
  return connection<RawUser, Result>(\\"user\\");
};

export function RawUserAccounts<Result = RawUserAccount[]>(
  connection: ConnectionOrTransaction
) {
  return connection<RawUserAccount, Result>(\\"user_account\\");
};

",
  "testCode": "import { RawUsers } from \\"./generated.ts\\"

describe(\\"RawUsers\\", () => {
  it(\\"doesn't throw\\", () => {
    expect(() => RawUsers(connection)).not.toThrow();
  });
});

import { RawUserAccounts } from \\"./generated.ts\\"

describe(\\"RawUserAccounts\\", () => {
  it(\\"doesn't throw\\", () => {
    expect(() => RawUserAccounts(connection)).not.toThrow();
  });
});

",
}
`;
